kernel/thread.c

```diff
@@ -348,40 +425,173 @@ long myst_syscall_wait4(
+                    p->znext->zprev = p->zprev;
+
+                if (_zombies_tail == p)
+                    _zombies_tail = p->group_prev;
+
+                _zombies_count--;
+
+                // remove from process list
+                myst_spin_lock(&myst_process_list_lock);
+                if (p->main.prev_process_thread)
+                    p->main.prev_process_thread->main.next_process_thread =
+                        p->main.next_process_thread;
+                if (p->main.next_process_thread)
+                    p->main.next_process_thread->main.prev_process_thread =
+                        p->main.prev_process_thread;
+                myst_spin_unlock(&myst_process_list_lock);
+
+                // free zombie thread
+                free(p);
+
                 goto done;
             }
         }
 
+        /* no zombies found, but are there any running childred? If not then
+         * ECHILD returned */
+        myst_spin_lock(&myst_process_list_lock);
+
+        // first processes left
+        p = process->main.prev_process_thread;
+        while (p && !_wait_matcher("active", pid, process, p))
+        {
+            p = p->main.prev_process_thread;
+        }
+        if (p == NULL)
+        {
+            // now processes right
+            p = process->main.next_process_thread;
+            while (p && !_wait_matcher("active", pid, process, p))
+            {
+                p = p->main.next_process_thread;
+            }
+        }
+        if (p == NULL)
+        {
+#ifdef TRACE
+            printf("Returning ECHILD\n");
+#endif
+            myst_spin_unlock(&myst_process_list_lock);
+            ERAISE(-ECHILD);
+        }
+
+        myst_spin_unlock(&myst_process_list_lock);
+
         if ((options & WNOHANG))
         {
             ret = 0;
             goto done;
         }
 
         /* wait for signal from myst_zombify_thread() */
-        myst_cond_wait(&_zombies_cond, &_zombies_mutex);
+#ifdef TRACE
+        if (pid == 0)
+        {
+            printf(
+                "Waiting for new zombification of child with our pgid=%u of "
+                "our process %u\n",
+                process->main.pgid,
+                process->pid);
+        }
+        else if (pid > 0)
+        {
+            printf(
+                "Waiting for new zombification of child whose pid=%d from our "
+                "process %u\n",
+                pid,
+                process->pid);
+        }
+        else if (pid == -1)
+        {
+            printf(
+                "Waiting for new zombification of any child of our process "
+                "%u\n",
+                process->pid);
+        }
+        else /* < -1 */
+        {
+            printf(
+                "Waiting for new zombification of child whose pgid=%d from our "
+                "process %u\n",
+                -pid,
+                process->pid);
+        }
+#endif
+        myst_mutex_unlock(&_zombies_mutex);
+        myst_sleep_msec(100);
```

Comment:
Ideally we should use conditional variables to wake up the waiter.

Reply:
I am not sure condition variables work here anyway which is why things broke in the first place. What would the condition variable be? We have zombies? Well we have multiple waits going on here. When it is signaled, do we catch it? Did we miss it? There are very major problems with any kind of signaling. We would have to have a condition variable connected to every process, so if any process is woken we wait on our own and then we know if it is signaled. In that case a zombie would need to signal all running processes.